src/Protocols/PacketStream/AsyncFifo.hs

```diff
@@ -0,0 +1,60 @@
+{-# LANGUAGE NoImplicitPrelude #-}
+
+{- |
+Provides `asyncFifoC` for crossing clock domains in the packet stream protocol.
+-}
+module Protocols.PacketStream.AsyncFifo (asyncFifoC) where
+
+import Data.Maybe.Extra (toMaybe)
+
+import Clash.Explicit.Prelude (asyncFIFOSynchronizer)
+import Clash.Prelude
+
+import Protocols.Internal (Circuit, fromSignals)
+import Protocols.PacketStream.Base
+
+{- | Asynchronous FIFO circuit that can be used to safely cross clock domains.
+Uses `Clash.Explicit.Prelude.asyncFIFOSynchronizer` internally.
+-}
+asyncFifoC ::
+  forall
+    (depth :: Nat)
+    (dataWidth :: Nat)
+    (wDom :: Domain)
+    (rDom :: Domain)
+    (metaType :: Type).
+  ( KnownDomain wDom
+  , KnownDomain rDom
+  , KnownNat depth
+  , 2 <= depth
+  , KnownNat dataWidth
+  , 1 <= dataWidth
+  , NFDataX metaType
+  ) =>
+  -- | 2^depth is the number of elements this component can store
+  SNat depth ->
+  -- | Clock signal in the write domain
+  Clock wDom ->
+  -- | Reset signal in the write domain
+  Reset wDom ->
+  -- | Enable signal in the write domain
+  Enable wDom ->
+  -- | Clock signal in the read domain
+  Clock rDom ->
+  -- | Reset signal in the read domain
+  Reset rDom ->
+  -- | Enable signal in the read domain
+  Enable rDom ->
+  Circuit (PacketStream wDom dataWidth metaType) (PacketStream rDom dataWidth metaType)
+asyncFifoC depth wClk wRst wEn rClk rRst rEn = fromSignals ckt
```

Comment:
This circuit behaves weird. I'm quite used to just put `enableGen` on my global enable lines. However, when I do that for something involving this asyncFifoC while the reset is asserted, I'll basically be writing garbage into the fifo. That's because the `PacketStream` does not have an inherent validity, and since the fifo won't be full, the ack will be asserted.

Reply:
But I guess it's not a problem as the write pointer won't make any progress.

Reply:
It does however depend on having a single reset for the entire domain, which is not mandatory.

Reply:
What do you mean `PacketStream` does not have inherent validity?

Reply:
There is no `_tvalid` like in AxiStream. I guess the `abort` could serve that function? Should we then mandate that abort must be asserted when the component is in reset?

Reply:
It's wrapped in a `Maybe`, So a `Nothing` means it's invalid. `tvalid` is modelled exactly like this in our clash-protocols axi stream protocol as well.

Reply:
Ah, I missed that.

Reply:
Apologies for the misunderstanding.